graylog2-server/src/main/java/org/graylog2/indexer/datanode/DatanodeMigrationLockServiceImpl.java

```diff
@@ -0,0 +1,127 @@
+/*
+ * Copyright (C) 2020 Graylog, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the Server Side Public License, version 1,
+ * as published by MongoDB, Inc.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * Server Side Public License for more details.
+ *
+ * You should have received a copy of the Server Side Public License
+ * along with this program. If not, see
+ * <http://www.mongodb.com/licensing/server-side-public-license>.
+ */
+package org.graylog2.indexer.datanode;
+
+import com.github.rholder.retry.Attempt;
+import com.github.rholder.retry.RetryException;
+import com.github.rholder.retry.RetryListener;
+import com.github.rholder.retry.RetryerBuilder;
+import com.github.rholder.retry.StopStrategies;
+import com.github.rholder.retry.WaitStrategies;
+import com.google.common.util.concurrent.ThreadFactoryBuilder;
+import jakarta.inject.Inject;
+import org.graylog2.cluster.lock.Lock;
+import org.graylog2.cluster.lock.LockService;
+import org.graylog2.indexer.IndexSet;
+import org.graylog2.plugin.Tools;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+
+public class DatanodeMigrationLockServiceImpl implements DatanodeMigrationLockService {
```

Comment:
I do not fully understand the code, so please forgive me this silly question, but how do we handle multi-node setup?
Can each GL node have its own migration lock service?
If yes, is it important that `activeLocks` collection contains all the locks?
If yes, how do we get locks created by another GL nodes appear in that collection?

Reply:
It seems that `activeLocks` collection is used just for extending the locks?
So maybe me previous questions are not important at all...

Reply:
Each GL node will have its own migration lock service. The locks are persisted in the mongodb, so they are distributed and block parallel usage of the index set across the whole cluster. The activeLocks collection is there only to keep the locally requested locks from expiring. Otherwise they'll automatically disappear from the mongo collection, as they have defined a TTL. The local `activeLocks` collection doesn't need to be distributed, each node is refreshing its own locks. In the moment the node disappears, these lock disappear as well, which is a good thing.

Reply:
Please check if you can use the `RefreshingLockService` to refresh the locks instead of implementing your own refresh system. :slightly_smiling_face:

Reply:
The problem with RefreshingLockService is that it can hold one lock only. I'd need one instance of the service for each lock. Instead of implementing own refresh system I'd need to implement a system for managing service instances, which seems even worse :-)